Explorer/Assets/DCL/WebRequests/Analytics/WebRequestsAnalyticsContainer.cs

```diff
@@ -5,23 +5,46 @@ namespace DCL.WebRequests.Analytics
 {
     public class WebRequestsAnalyticsContainer : IWebRequestsAnalyticsContainer
     {
-        private readonly Dictionary<Type, int> ongoingRequests = new ();
+        private readonly Dictionary<Type, List<IRequestMetric>> requestTypesWithMetrics = new ();
+        private readonly List<Type> requestMetricTypes = new ();
+        public IReadOnlyList<IRequestMetric> GetMetric(Type requestType) =>
+            requestTypesWithMetrics.GetValueOrDefault(requestType);
 
-        public int GetMetric(Type requestType) =>
-            ongoingRequests.TryGetValue(requestType, out int count) ? count : 0;
+        public List<Type> GetTrackedMetrics() => requestMetricTypes;
 
+        public WebRequestsAnalyticsContainer AddTrackedMetric<T>() where T: IRequestMetric
+        {
+            requestMetricTypes.Add(typeof(T));
+            return this;
+        }
         void IWebRequestsAnalyticsContainer.OnRequestStarted<T>(T request)
         {
-            if (ongoingRequests.TryGetValue(typeof(T), out int count))
-                ongoingRequests[typeof(T)] = count + 1;
-            else
-                ongoingRequests.Add(typeof(T), 1);
+            if (!requestTypesWithMetrics.TryGetValue(typeof(T), out List<IRequestMetric> metrics))
+            {
+                metrics = new List<IRequestMetric>();
+
+                foreach (Type metricType in requestMetricTypes)
+                {
+                    metrics.Add((IRequestMetric)Activator.CreateInstance(metricType));
```

Comment:
Instead of using reflection, you could provide a delegate in `AddTrackedMetric` method with appropriate constraints.
Even if you think it's not needed at least provide the `new()` constraint to ensure that the instance of the type can be created by `Activator`

Reply:
If you provide a delegate you don't need a separate `requestMetricTypes` collection as you can construct everything in `AddTrackedMateric` and then expose its keys. Another argument in favor of passing a delegate

Reply:
As discussed, we cannot use delegates as we need to store the type because of the lazy creation